benchmarks/cpp/nvfuser/transpose.cpp

```diff
@@ -0,0 +1,260 @@
+    // ->RangeMultiplier(2)
+    ->Ranges({{4, 1024 * 1024}, {160, 320}})
+    ->Unit(benchmark::kMicrosecond)
+    ->UseManualTime();
+
+//------------------------------------------------------------------------------
+
+static void Baseline_Transpose(
+    benchmark::State& benchmark_state,
+    DataType dtype,
+    int num_dims,
+    std::pair<int, int> axes,
+    bool input1_transpose_axes = false,
+    bool input2_transpose_axes = false,
+    bool intermediate_transpose_axes = false,
+    bool output_transpose_axes = false) {
+  auto aten_inputs = generateInputs(
+      dtype,
+      num_dims,
+      axes,
+      benchmark_state.range(0),
+      benchmark_state.range(1),
+      input1_transpose_axes,
+      input2_transpose_axes);
+  auto at_input1 = aten_inputs[0];
+  auto at_input2 = aten_inputs[1];
+
+  auto optionalTransposeAten = [&axes](at::Tensor at, bool is_transpose) {
+    return (is_transpose) ? at::transpose(at, axes.first, axes.second) : at;
+  };
+
+  for (auto _ : benchmark_state) {
+    clearL2Cache();
+    CudaKernelTimer timer;
+
+    auto at_ot_input1 =
+        optionalTransposeAten(at_input1, input1_transpose_axes);
+    auto at_ot_input2 =
+        optionalTransposeAten(at_input2, input2_transpose_axes);
+    auto at_intermediate = add(at_ot_input1, at_ot_input2);
+    auto at_ot_intermediate =
+        optionalTransposeAten(at_intermediate, intermediate_transpose_axes);
+    auto at_output = relu(at_ot_intermediate);
+    auto at_ot_output =
+        optionalTransposeAten(at_output, output_transpose_axes);
+
+    benchmark_state.SetIterationTime(timer.elapsed() / 1000.0);
+  }
+  // Sync everything up before we're finished, don't want to run ahead on the
+  // cpu while benchmarking.
+  cudaDeviceSynchronize();
+
+  benchmark_state.SetBytesProcessed(
+      int64_t(benchmark_state.iterations()) *
+      (at_input1.numel() * 3 * int64_t(dataTypeSize(dtype))));
+}
+
+static void Baseline_Transpose_Inner_fp32(benchmark::State& benchmark_state) {
+  Baseline_Transpose(
+      benchmark_state,
+      DataType::Float,
+      2 /* num_dims */,
+      {0, 1} /* axes */,
+      true /* input1_transpose_axes */,
+      false /* input2_transpose_axes */,
+      false /* intermediate_transpose_axes */,
+      false /* output_transpose_axes */);
+}
+
+static void Baseline_Transpose_Inner_fp16(benchmark::State& benchmark_state) {
+  Baseline_Transpose(
+      benchmark_state,
+      DataType::Half,
+      2 /* num_dims */,
+      {0, 1} /* axes */,
+      true /* input1_transpose_axes */,
+      false /* input2_transpose_axes */,
+      false /* intermediate_transpose_axes */,
+      false /* output_transpose_axes */);
+}
+
+//------------------------------------------------------------------------------
+
+BENCHMARK(Baseline_Transpose_Inner_fp32)
+    // ->RangeMultiplier(2)
+    ->Ranges({{4, 1024 * 1024}, {160, 320}})
+    ->Unit(benchmark::kMicrosecond)
+    ->UseManualTime();
+
+BENCHMARK(Baseline_Transpose_Inner_fp16)
```

Comment:
I add 3D tensor benchmarks and more square matrix shape variations. I may refactor some of the macros to avoid creating all 64 transpose combinations directly.